plugins/obs-ffmpeg/obs-ffmpeg-hls-mux.c

```diff
@@ -0,0 +1,336 @@
+{
+	circlebuf_push_back(&stream->packets, packet,
+			    sizeof(struct encoder_packet));
+	return true;
+}
+
+static void drop_frames(struct ffmpeg_muxer *stream, int highest_priority)
+{
+	struct circlebuf new_buf = {0};
+	int num_frames_dropped = 0;
+
+	circlebuf_reserve(&new_buf, sizeof(struct encoder_packet) * 8);
+
+	while (stream->packets.size) {
+		struct encoder_packet packet;
+		circlebuf_pop_front(&stream->packets, &packet, sizeof(packet));
+
+		/* do not drop audio data or video keyframes */
+		if (packet.type == OBS_ENCODER_AUDIO ||
+		    packet.drop_priority >= highest_priority) {
+			circlebuf_push_back(&new_buf, &packet, sizeof(packet));
+		} else {
+			num_frames_dropped++;
+			obs_encoder_packet_release(&packet);
+		}
+	}
+
+	circlebuf_free(&stream->packets);
+	stream->packets = new_buf;
+
+	if (stream->min_priority < highest_priority)
+		stream->min_priority = highest_priority;
+
+	stream->dropped_frames += num_frames_dropped;
+}
+
+static bool find_first_video_packet(struct ffmpeg_muxer *stream,
+				    struct encoder_packet *first)
+{
+	size_t count = stream->packets.size / sizeof(*first);
+
+	for (size_t i = 0; i < count; i++) {
+		struct encoder_packet *cur =
+			circlebuf_data(&stream->packets, i * sizeof(*first));
+		if (cur->type == OBS_ENCODER_VIDEO && !cur->keyframe) {
+			*first = *cur;
+			return true;
+		}
+	}
+	return false;
+}
+
+void check_to_drop_frames(struct ffmpeg_muxer *stream, bool pframes)
+{
+	struct encoder_packet first;
+	int64_t buffer_duration_usec;
+	int priority = pframes ? OBS_NAL_PRIORITY_HIGHEST
+			       : OBS_NAL_PRIORITY_HIGH;
+	int keyint_sec = stream->keyint_sec;
+	int64_t drop_threshold_sec = keyint_sec ? 2 * keyint_sec : 10;
+
+	if (!find_first_video_packet(stream, &first))
+		return;
+
+	buffer_duration_usec = stream->last_dts_usec - first.dts_usec;
+
+	if (buffer_duration_usec > drop_threshold_sec * 1000000)
+		drop_frames(stream, priority);
+}
+
+static bool add_video_packet(struct ffmpeg_muxer *stream,
+			     struct encoder_packet *packet)
+{
+	check_to_drop_frames(stream, false);
+	check_to_drop_frames(stream, true);
+
+	/* if currently dropping frames, drop packets until it reaches the
+	 * desired priority */
+	if (packet->drop_priority < stream->min_priority) {
+		stream->dropped_frames++;
+		return false;
+	} else {
+		stream->min_priority = 0;
+	}
+
+	stream->last_dts_usec = packet->dts_usec;
+	return write_packet_to_buf(stream, packet);
+}
+
+void ffmpeg_hls_mux_data(void *data, struct encoder_packet *packet)
+{
+	struct ffmpeg_muxer *stream = data;
+	struct encoder_packet new_packet;
+	struct encoder_packet tmp_packet;
+	bool added_packet = false;
+
+	if (!active(stream))
+		return;
+
+	/* encoder failure */
+	if (!packet) {
+		deactivate(stream, OBS_OUTPUT_ENCODE_ERROR);
+		return;
+	}
+
+	if (!stream->sent_headers) {
+		if (!send_headers(stream))
+			return;
+		stream->sent_headers = true;
+	}
+
+	if (stopping(stream)) {
+		if (packet->sys_dts_usec >= stream->stop_ts) {
```

Comment:
I'm wondering if sys_dts_usec is actually set. Could you log this value and check?